### PR TITLE
Clean Up trace pkg

### LINKED ISSUE #145: Clean up trace package
## Is your feature request related to a problem? Please describe

The trace package ahs code that should live in the respective package.
Example: `SQSConsumerComponent` constant should live in the SQS consumer package.
We should keep the trace package clean and implementation agnostic.

### PR DESCRIPTION
**Pull request by tpaschalis**
## Which problem is this PR solving?

This PR closes #145 [Clean up trace package](https://github.com/beatlabs/patron/issues/145). 

The end-goal is to refactor code which should live in another packages, and keep the base `trace/` package as lean and implementation-agnostic as possible.

## Short description of the changes

* Break up AMQP/Kafka/SQS Consumer component constant names in the `async/` sub-package.
* Break up AMQP/Kafka/SNS Publisher component names to the respective `trace/` sub-package.
* Break up the HTTP and SQL component names to the destination sub-package.
* For SQL, HTTP, and ElasticSearc, SQLMove the `XYZSpan()` function into their respective sub-package. Rename them to avoid stuttering. If a `startSpan()` function exists, integrate the functionality there, so there's exactly one entrypoint.
* Make `hostsTag` and `versionTag` public so they're re-used between current and future implementations.
* Fix and distribute unittests in their respective packages.
* TODO Check one more time on whether we can unexport some of the functions.

## Notes
The HTTP span functions have been left unexported, as they're used elsewhere (eg. in the middleware layer, and in the elasticsearch package). They're generic enough to be used in future places too, so I think they should remain exported, rather than copied.

### CODE REVIEW

**Review by drakos74 — COMMENTED** [review]
It's not clear to me what the collective meaning of "sync" , "async" and "trace" .
I would imagine a logical separation at this level.
Sync and async fit together, but "trace" is a bit outside of their scope.
Would it make sense to use consumer / producer as another granularity level for the packages ?
Apart from that, changes look good to me. And refactoring / reorganising the packages definitely makes sense!

File: async/amqp/amqp.go [hunk 1]
```diff
@@ -18,6 +18,11 @@ import (
 	"github.com/streadway/amqp"
 )
 
+const (
+	// AMQPConsumerComponent definition.
+	AMQPConsumerComponent = "amqp-consumer"
```
Comment (mantzas) [inline async/amqp/amqp.go]: if this is not used outside of the package we can unexport it and remove the AMQP prefix since it already lives in the amqp package

File: async/kafka/kafka.go [hunk 2]
```diff
@@ -18,6 +18,11 @@ import (
 	"github.com/prometheus/client_golang/prometheus"
 )
 
+const (
+	// KafkaConsumerComponent definition.
+	KafkaConsumerComponent = "kafka-consumer"
```
Comment (mantzas) [inline async/kafka/kafka.go]: same as above: unexport and remove prefix

File: async/sqs/sqs.go [hunk 3]
```diff
@@ -34,6 +34,9 @@ const (
 	ackMessageState     messageState = "ACK"
 	nackMessageState    messageState = "NACK"
 	fetchedMessageState messageState = "FETCHED"
+
+	// SQSConsumerComponent definition.
+	SQSConsumerComponent = "sqs-consumer"
```
Comment (mantzas) [inline async/sqs/sqs.go]: smae as above

File: sync/http/middleware.go [hunk 4]
```diff
@@ -8,7 +8,7 @@ import (
 	"github.com/beatlabs/patron/correlation"
 	"github.com/beatlabs/patron/log"
 	"github.com/beatlabs/patron/sync/http/auth"
-	"github.com/beatlabs/patron/trace"
+	trace_http "github.com/beatlabs/patron/trace/http"
```
Comment (mantzas) [inline sync/http/middleware.go:11]: why this change ? did we have an import collision?

File: trace/amqp/amqp.go [hunk 5]
```diff
@@ -17,6 +17,11 @@ import (
 	"github.com/streadway/amqp"
 )
 
+const (
+	// AMQPPublisherComponent definition.
+	AMQPPublisherComponent = "amqp-publisher"
```
Comment (mantzas) [inline trace/amqp/amqp.go]: same as before

File: trace/http/http.go [hunk 6]
```diff
@@ -6,13 +6,21 @@ import (
...
 	"github.com/beatlabs/patron/reliability/circuitbreaker"
 	"github.com/beatlabs/patron/trace"
 	"github.com/opentracing-contrib/go-stdlib/nethttp"
 	opentracing "github.com/opentracing/opentracing-go"
 	"github.com/opentracing/opentracing-go/ext"
 )
 
+const (
+	// HTTPComponent definition.
+	HTTPComponent = "http"
```
Comment (mantzas) [inline trace/http/http.go]: need to be exported? prefix needed?

File: trace/kafka/kafka.go [hunk 7]
```diff
@@ -14,6 +14,11 @@ import (
 	"github.com/opentracing/opentracing-go/ext"
 )
 
+const (
+	// KafkaAsyncProducerComponent definition.
+	KafkaAsyncProducerComponent = "kafka-async-producer"
```
Comment (mantzas) [inline trace/kafka/kafka.go]: export, prefix?

File: trace/sns/publisher.go [hunk 8]
```diff
@@ -14,6 +14,11 @@ import (
 	"github.com/opentracing/opentracing-go/ext"
 )
 
+const (
+	// SNSPublisherComponent definition.
+	SNSPublisherComponent = "sns-publisher"
```
Comment (mantzas) [inline trace/sns/publisher.go]: export? prefix?

**Review by mantzas — CHANGES_REQUESTED** [review]

**Comment by mantzas** [comment]
> It's not clear to me what the collective meaning of "sync" , "async" and "trace" .
> I would imagine a logical separation at this level.
> Sync and async fit together, but "trace" is a bit outside of their scope.
> Would it make sense to use consumer / producer as another granularity level for the packages ?
> Apart from that, changes look good to me. And refactoring / reorganising the packages definitely makes sense!

@drakos74 you are right we have an open ticket for that... #137

File: async/amqp/amqp.go (hunk 1 above)
Comment (tpaschalis) [reply async/amqp/amqp.go]: @mantzas regarding this and the other comments below; I agree with removing the prefix to avoid stuttering. 

On the other hand, I think we should keep it exported. This way people using these packages will have a uniform way available to refer to the component names (eg. while logging) which will correspond to the name used in tracing.

File: sync/http/middleware.go (hunk 4 above)
Comment (tpaschalis) [reply sync/http/middleware.go:11]: @mantzas The base trace package is not used anymore, as functionality has moved to trace/http, which collides with net/http.

File: async/amqp/amqp.go (hunk 1 above)
Comment (tpaschalis) [reply async/amqp/amqp.go]: @mantzas any thoughts?

Comment (mantzas) [reply async/amqp/amqp.go]: this is actually internal and should probably not be used outside the package. We can always export it if somebody wants it for some reason.

**Review by mantzas — APPROVED** [review]

**Review by sotirispl — APPROVED** [review]

**Review by drakos74 — APPROVED** [review]
